metrics/tracer.go

```diff
@@ -0,0 +1,142 @@
+package metrics
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/quic-go/quic-go/internal/protocol"
+	"github.com/quic-go/quic-go/internal/qerr"
+	"github.com/quic-go/quic-go/logging"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const metricNamespace = "quicgo"
+
+func getIPVersion(addr net.Addr) string {
+	udpAddr, ok := addr.(*net.UDPAddr)
+	if !ok {
+		return ""
+	}
+	if udpAddr.IP.To4() != nil {
+		return "ipv4"
+	}
+	return "ipv6"
+}
+
+var (
+	connsRejected = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNamespace,
+			Name:      "server_connections_rejected_total",
+			Help:      "Connections Rejected",
+		},
+		[]string{"ip_version", "reason"},
+	)
+	packetDropped = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNamespace,
+			Name:      "server_received_packets_dropped_total",
+			Help:      "packets dropped",
+		},
+		[]string{"ip_version", "reason"},
+	)
+)
+
+// NewTracer creates a new tracer using the default Prometheus registerer.
+// The Tracer returned from this function can be used to collect metrics for
+// events happening before the establishment of a QUIC connection.
+// It can be set on the Tracer field of quic.Transport.
+func NewTracer() *logging.Tracer {
+	return NewTracerWithRegisterer(prometheus.DefaultRegisterer)
+}
+
+// NewTracerWithRegisterer creates a new tracer using a given Prometheus registerer.
+func NewTracerWithRegisterer(registerer prometheus.Registerer) *logging.Tracer {
+	for _, c := range [...]prometheus.Collector{
+		connsRejected,
+		packetDropped,
+	} {
+		if err := registerer.Register(c); err != nil {
+			if ok := errors.As(err, &prometheus.AlreadyRegisteredError{}); !ok {
+				panic(err)
+			}
+		}
+	}
+
+	return &logging.Tracer{
+		SentPacket: func(addr net.Addr, hdr *logging.Header, _ logging.ByteCount, frames []logging.Frame) {
+			tags := getStringSlice()
+			defer putStringSlice(tags)
+
+			var reason string
+			switch {
+			case hdr.Type == protocol.PacketTypeRetry:
+				reason = "retry"
+			case hdr.Type == protocol.PacketTypeInitial:
+				var ccf *logging.ConnectionCloseFrame
+				for _, f := range frames {
+					cc, ok := f.(*logging.ConnectionCloseFrame)
+					if ok {
+						ccf = cc
+						break
+					}
+				}
+				if ccf.IsApplicationError {
```

Comment:
Do we need a `ccf != nil` here?

Reply:
Technically not, since the only reason we send an Initial packet before creating a connection is struct is to reject the connection attempt. We should document that though.

Reply:
Or rather, be more defensive here. It's not good to panic just to collect metrics.